File: BackTracking/N-Queens.py
def solveNQueens(N):
    def is_safe(board, row, col):
        for i in range(row):
            if board[i][col] == 'Q':
                return False
            #checking for upper-left diagonal

            i, j = row - 1, col - 1
            while i >= 0 and j >= 0:
                if board[i][j] == 'Q':
                    return False
                j -= 1
                i -= 1
            
            #checks for upper-right diagonal
            i, j = row - 1, col + 1
            while i >= 0 and j < N:
                if board[i][j] == 'Q':
                    return False
                i -= 1
                j += 1

        return True
    
    def backtrack(board, row):
        if row == N:
            solutions.append([''.join(r) for r in board])
            return
        
        for col in range(N):
            if is_safe(board, row, col):
                board[row][col] = 'Q'
                backtrack(board, row + 1)
                board[row][col] = '.'
    board = [['.' for _ in range(N)] for _ in range(N)]
    solutions = []
    backtrack(board, 0)
    return solutions
    
N = 10
solutions = solveNQueens(N)

print(solutions)

for solution in solutions:
    for key in solution:
        print(key)
    print()
        

File: algorithm-analysis/exam/back-tracking/n-queens.c
#include <stdio.h>

#define N 5

int isSafe(int board[N][N],int row, int col) {
    int flag = 1;
    for (int i = 0; i < row; i++) {
        if (board[i][col] == 1) {
            flag = 0;
        }
    }

    for (int i = row - 1, j = col - 1; i >= 0 && j >= 0; i--, j--) {
        if (board[i][j] == 1) {
            flag = 0;
        }
    }

    for (int i = row - 1, j = col + 1; i >= 0 && j < N; i--, j++) {
        if (board[i][j] == 1) {
            flag = 0;
        }
    }

    return flag;
}

void solveNQueens(int board[N][N], int row) {
    if (row == N) {
        printf("Solution: \n");
        for (int i = 0; i < N; i++)
        {
            for (int j = 0; j < N; j++)
            {
                printf("%d\t", board[i][j]);
            }
            printf("\n");
        }
        return;
    }
    for (int i = 0; i < N; i++) {
        if (isSafe(board, row, i)) {
            board[row][i] = 1;
            solveNQueens(board, row + 1);
            board[row][i] = 0;
        }

    }
}

int main()
{
    int board[N][N];

    for (int i = 0; i < N; i++)
    {
        for (int j = 0; j < N; j++)
        {
            board[i][j] = 0;
        }
    }

    solveNQueens(board, 0);

    return 0;
}